### PR TITLE
Added support to Quartz Plugins And Listeners

### LINKED ISSUE #10962: add a way to configure "plugins" and "listeners" for the quartz scheduler
**Description**

Quartz has two concepts of listeners and plugins which I think can be useful to add things like Logging, Tracking job execution history in an event store, database etc: 

Example of the common used plugin is the [LoggingHistoryPlugin](https://github.com/quartz-scheduler/quartz/blob/d42fb7770f287afbf91f6629d90e7698761ad7d8/quartz-plugins/src/main/java/org/quartz/plugins/history/LoggingJobHistoryPlugin.java)

**Implementation ideas**

Add the two entries i.e "listeners" and "plugins" in [QuartzBuildTimeConfig](https://github.com/quarkusio/quarkus/blob/ac788aafbf9f01908aa379a29d91d0d67dec36f9/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzBuildTimeConfig.java). Their type can be a "Map<String, Map<String, String>>", allowing for customization of different properties a listener / plugin can have.  

We just have to be add a validation that the "class" key is there in the second Map cause this is what will define the listener / plugin class which will have to be registered for reflection so that it works in native mode.

The configuration will need to be mapped to corresponding Quartz configuration see https://github.com/quarkusio/quarkus/blob/ac788aafbf9f01908aa379a29d91d0d67dec36f9/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L288. 

See Quartz listener / plugin configuration references:
- http://www.quartz-scheduler.org/documentation/quartz-2.2.2/configuration/ConfigListeners.html
- http://www.quartz-scheduler.org/documentation/quartz-2.3.0/configuration/ConfigPlugins.html

/cc @mkouba

### PR DESCRIPTION
**Pull request by marcelorubim**
Added support to Quartz Plugins And Listeners, as described on issue #10962 

Fix #10962

### CODE REVIEW

**Comment by marcelorubim** [comment]
Well, there is no log at the failed pipe (Quarkus CI / JDK Java 14 JVM Tests (pull_request)). Should I send another commit to it to retry?

**Comment by mkouba** [comment]
Hi @marcelorubim, thanks for your contribution! Your PR should include at least some basic tests (you can follow the tests in https://github.com/quarkusio/quarkus/tree/master/extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test). Furthermore, your branch should be squashed down to a single buildable commit. And finally, the [issue](https://github.com/quarkusio/quarkus/issues/10962) is still being discussed so pls keep an eye on the comments...

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/RegisterShutdownHookPluginTest.java
Comment (mkouba) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/RegisterShutdownHookPluginTest.java:27]: That's not a good plugin to test. We do shutdown the quartz scheduler when a quarkus app/test stops ;-).

Comment (marcelorubim) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/RegisterShutdownHookPluginTest.java]: This test is because if I register an invalid plugin, the Quartz won't start. The test can check if parsing the config to Quartz properties is working fine.

**Review by machi1990 — CHANGES_REQUESTED** [review]
Thank you for the PR @marcelorubim I'd suggest we test we add an integration test too. Things like [class registration for reflection](https://github.com/quarkusio/quarkus/pull/10967/files#diff-9aa5c80fb75e537c6b40123ab4ac797bR16) would most likely be needed and having an integration test https://github.com/quarkusio/quarkus/tree/master/integration-tests/quartz would most likely show this up and avoid future regressions.

**Comment by marcelorubim** [comment]
> Thank you for the PR @marcelorubim I'd suggest we test we add an integration test too. Things like [class registration for reflection](https://github.com/quarkusio/quarkus/pull/10967/files#diff-9aa5c80fb75e537c6b40123ab4ac797bR16) would most likely be needed and having an integration test https://github.com/quarkusio/quarkus/tree/master/integration-tests/quartz would most likely show this up and avoid future regressions.

Added this on application.properties on the integration test:

```
# Register de LoggingJobHistoryPlugin for testing
quarkus.quartz.plugin.jobHistory.class=org.quartz.plugins.history.LoggingJobHistoryPlugin
quarkus.quartz.plugin.jobHistory.jobSuccessMessage=Job [{1}.{0}] execution complete and reports: {8}
```

And this to QuartzProcessor. 

```java
        config.triggerListeners.values().forEach((value) -> {
            reflectiveClasses.add(new ReflectiveClassBuildItem(true, false, value.clazz));
        });

        config.jobListeners.values().forEach((value) -> {
            reflectiveClasses.add(new ReflectiveClassBuildItem(true, false, value.clazz));
        });

        config.plugins.values().forEach((value) -> {
            reflectiveClasses.add(new ReflectiveClassBuildItem(true, false, value.clazz));
        });
```

Now seems to works just fine on native mode.

File: extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java
Comment (machi1990) [inline extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java:149]: I think we should also add validation here that the `class` is a class implementing a `org.quartz.JobListener` interface. This way, we would throw the error earlier in build time. Let me know if you'd need help in doing it.

/cc @mkouba

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (machi1990) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:346]: We tend to avoid usage of lambdas in runtime modules. Can we go for classic for loops here? Probably extracting this logic into a method `addListenerProperty(String prefix, QuartzAdditionalPropsConfig config) {...}`, where the prefix is one of 
- `StdSchedulerFactory.PROP_PLUGIN_PREFIX` , 
- `StdSchedulerFactory.PROP_JOB_LISTENER_PREFIX` and 
-`StdSchedulerFactory.PROP_JOB_LISTENER_PREFIX` 

would remove some of the duplication.

File: integration-tests/quartz/src/main/resources/application.properties
Comment (machi1990) [inline integration-tests/quartz/src/main/resources/application.properties:21]: Nice. 

Do you think we can be able to add a line or two in the [Quartz guide](https://github.com/quarkusio/quarkus/blob/ed51c0e624d1e89ebaf9ee6f1fd971e8056689b9/docs/src/main/asciidoc/quartz.adoc) about configuring listeners and plugins?

**Review by machi1990 — COMMENTED** [review]
@marcelorubim thank you for the wonderful work. It is looking great, I like your configuration proposal much better than the plain `Map<String, Map<String, String>>`. 

I've added comments, and once they are addressed this is good to merge.

File: integration-tests/quartz/src/main/resources/application.properties
Comment (machi1990) [reply integration-tests/quartz/src/main/resources/application.properties:21]: See https://github.com/quarkusio/quarkus/issues/10962#issuecomment-664231324

File: extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java
Comment (machi1990) [reply extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java:149]: See https://github.com/quarkusio/quarkus/issues/10962#issuecomment-664231324

Comment (marcelorubim) [reply extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java]: Thanks for the feedback! I did the changes you asked.
For the validation, the only thing I was not sure of which the exception I should throw. I decided on the IllegalArgumentException. Is it ok?

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (marcelorubim) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: I did the refactoring. It's much better this way, without the duplication

```java
        props.putAll(getAdditionalConfigurationProperties(StdSchedulerFactory.PROP_PLUGIN_PREFIX, buildTimeConfig.plugins));
        props.putAll(getAdditionalConfigurationProperties(StdSchedulerFactory.PROP_JOB_LISTENER_PREFIX,
                buildTimeConfig.jobListeners));
        props.putAll(getAdditionalConfigurationProperties(StdSchedulerFactory.PROP_TRIGGER_LISTENER_PREFIX,
                buildTimeConfig.triggerListeners));

        return props;
    }

    private Properties getAdditionalConfigurationProperties(String prefix, Map<String, QuartzAdditionalPropsConfig> config) {
        Properties props = new Properties();
        for (String key : config.keySet()) {
            props.put(String.format("%s.%s.class", prefix, key), config.get(key).clazz);
            for (String propsName : config.get(key).propsValue.keySet()) {
                props.put(String.format("%s.%s.%s", prefix, key, propsName),
                        config.get(key).propsValue.get(propsName));
            }
        }
        return props;
    }
```

File: docs/src/main/asciidoc/quartz.adoc
Comment (machi1990) [inline docs/src/main/asciidoc/quartz.adoc:416]: This is much better. Thank you. 

 Also in this issue we wanted to document the programmatic way of doing things as pointed out by @mkouba in https://github.com/quarkusio/quarkus/issues/10962#issuecomment-664152333. Can you add a note in the documentation about that? Thank you.

File: extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java
Comment (machi1990) [reply extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java]: This is fine for me. 

I am thinking maybe we can have a unit tests for validation. The [UnsupportedClusteredJobConfigurationTest.java](https://github.com/quarkusio/quarkus/blob/master/extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/UnsupportedClusteredJobConfigurationTest.java) can be of inspiration.  

What do you think?

Comment (marcelorubim) [reply extensions/quartz/deployment/src/main/java/io/quarkus/quartz/deployment/QuartzProcessor.java]: Nice idea! And done!

File: docs/src/main/asciidoc/quartz.adoc
Comment (marcelorubim) [reply docs/src/main/asciidoc/quartz.adoc]: I included this information on the documentation. But I didn't find on Quartz API where to register a Plugin programmatic. So I just pointed out this way for jobListener or triggerListener.

Do you know if it is possible?

Comment (machi1990) [reply docs/src/main/asciidoc/quartz.adoc]: Nice work. I do not know if it possible.

**Review by machi1990 — APPROVED** [review]
Thank you @marcelorubim. All good for me. 


@mkouba do you want to have a look?

File: docs/src/main/asciidoc/quartz.adoc
Comment (mkouba) [inline docs/src/main/asciidoc/quartz.adoc:408]: _The example bellow register**s** the plugin..._

File: docs/src/main/asciidoc/quartz.adoc
Comment (mkouba) [inline docs/src/main/asciidoc/quartz.adoc:416]: You can also register a listener programmatically with an injected `org.quartz.Scheduler`...

**Review by mkouba — COMMENTED** [review]
Looks really good. I've added few minor docs comments.

File: docs/src/main/asciidoc/quartz.adoc
Comment (marcelorubim) [reply docs/src/main/asciidoc/quartz.adoc]: Thanks! I fixed it!

File: docs/src/main/asciidoc/quartz.adoc
Comment (marcelorubim) [reply docs/src/main/asciidoc/quartz.adoc]: Done! Really seems better this way!

**Review by mkouba — APPROVED** [review]

File: docs/src/main/asciidoc/quartz.adoc
Comment (marcelorubim) [reply docs/src/main/asciidoc/quartz.adoc]: Seems something is wrong with the CI itself.
The previous run was successful and on this one, I've only changed the docs.

/cc @machi1990

Comment (machi1990) [reply docs/src/main/asciidoc/quartz.adoc]: It could be only a case of failing random test. Can you rebase on master and then push again to re-launch CI?

Comment (marcelorubim) [reply docs/src/main/asciidoc/quartz.adoc]: I'll do that. Thanks!

**Review by gsmet — CHANGES_REQUESTED** [review]
Minor gripe: any chance you could adjust the commit message? `git commit --amend` should allow you to do that and then force push.

Thanks!

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks! Let's wait for CI and merge.

**Comment by gsmet** [comment]
Merged, thanks!

**Comment by machi1990** [comment]
Thank you @marcelorubim for the PR.

**Pull request merged**
